src/Control/Monad/Gen.purs

```diff
@@ -118,22 +119,12 @@ filtered gen = tailRecM go unit
     Nothing -> Loop unit
     Just a' -> Done a'
 
--- | Internal: used by fromIndex
-newtype AtIndex a = AtIndex (Int -> a)
-
-instance semigroupAtIndex :: Semigroup (AtIndex a)
-  where
-  append (AtIndex f) (AtIndex g) =
-    AtIndex \i -> if i <= 0 then f i else g (i - 1)
-
-atIndex :: forall a. a -> AtIndex a
-atIndex = AtIndex <<< const
-
-getAtIndex :: forall a. AtIndex a -> Int -> a
-getAtIndex (AtIndex f) = f
-
 -- | Internal: get the Foldable element at index i.
 -- | If the index is <= 0, return the first element.
 -- | If it's >= length, return the last.
 fromIndex :: forall f a. Foldable1 f => Int -> f a -> a
-fromIndex i xs = getAtIndex (foldMap1 atIndex xs) i
+fromIndex i xs = go i (foldr Cons Nil xs)
+  where
+    go j (Cons a _) | j <= 0 = a
+    go j (Cons _ as) = go (j - 1) as
+    go _ Nil = un Last (foldMap1 Last xs)
```

Comment:
If you added a case for `Cons a Nil` you could avoid another traversal, and instead this would just serve as proof of non-emptyness, but never actually be invoked.